crates/cranelift/src/compiler/component.rs

```diff
@@ -262,7 +258,163 @@ impl<'a> TrampolineCompiler<'a> {
                 let params = self.builder.func.dfg.block_params(self.block0);
                 (params[2], params[3])
             }
-        };
+        }
+    }
+
+    fn translate_task_return_call(&mut self) {
+        match self.abi {
+            Abi::Wasm => {}
+
+            // These trampolines can only actually be called by Wasm, so
+            // let's assert that here.
+            Abi::Array => {
+                self.builder.ins().trap(TRAP_INTERNAL_ASSERT);
+                return;
+            }
+        }
+
+        let args = self.builder.func.dfg.block_params(self.block0).to_vec();
+        let vmctx = args[0];
+
+        let (values_vec_ptr, values_vec_len) = self.store_wasm_arguments(&args[2..]);
+
+        let (host_sig, index) = host::task_return(self.isa, &mut self.builder.func);
+        let host_fn = self.load_libcall(vmctx, index);
+
+        let params = self.types[self.signature]
+            .unwrap_func()
+            .params()
+            .iter()
+            .map(|&v| {
+                Some(match v {
+                    WasmValType::I32 => FlatType::I32,
+                    WasmValType::I64 => FlatType::I64,
+                    WasmValType::F32 => FlatType::F32,
+                    WasmValType::F64 => FlatType::F64,
+                    _ => return None,
+                })
+            })
+            .collect::<Option<_>>();
+
+        let ty = self.builder.ins().iconst(
+            ir::types::I32,
+            i64::from(
+                params
+                    .and_then(|params| {
+                        self.types
+                            .get_task_return_type(&TypeTaskReturn { params })
+                            .map(|v| v.as_u32())
+                    })
+                    .unwrap_or(u32::MAX),
+            ),
+        );
```

Comment:
Two things on this:

* For `let params = ...` above, it's unclear to me what this is doing. For example I'm not sure why this would switch to `None` for unknown types as opposed to return an error or panic or something. Additionally this looks like something that's better done in perhaps a `translate.rs` or `inline.rs` step? 
* For `let ty = ...` this looks like it's doing a hash map lookup of a calculated type. Would it be possible to calculate that when the type is originally created early on in `translate.rs` and/or `inline.rs` and then thread that through to here? For example I'd expect that this function here would have `ty: Option<SomeIndex>` as an argument and then that's what's encoded into a `u32` here.

Reply:
Agreed about the `None` default; I don't remember why I did that.  I'll switch it to a panic.

And yes, I can move this code to an earlier stage, e.g. `translate.rs` or `inline.rs`.

Reply:
I just pushed an update which uses https://github.com/bytecodealliance/wasm-tools/pull/1989 and makes this code obsolete.  Now we compare `TypeTupleIndex`es, so no need to intern core function types anymore.